torch/nn/parallel/distributed.py

```diff
@@ -315,6 +297,22 @@ def _ddp_init_helper(self):
         (5) passing a handle of DDP to SyncBatchNorm Layer
         """
         if self.device_ids and len(self.device_ids) > 1:
+
+            import warnings
+            warnings.warn(
+                "Single-Process Multi-GPU is not the recommended mode for "
+                "DDP. In this mode, each DDP instance operates on multiple "
+                "devices and creates multiple module replicas within one "
+                "process. The overhead of scatter/gather and GIL contention "
+                "in every forward pass can slow down training. "
+                "Please consider using one DDP instance per device or per "
+                "module replica by explicitly setting device_ids or "
+                "CUDA_VISIBLE_DEVICES. "
+                "NB: There is a known issue in nn.parallel.replicate that "
```

Comment:
nit (not necessary to fix): it would be good to direct to a specific github issue where users can learn more and/or see if it's since been resolved.